### PR TITLE
NIFI-11158 PutSalesforceObject processor improvements

### PR DESCRIPTION
**Pull request by krisztina-zsihovszki**
# Summary

[NIFI-11158](https://issues.apache.org/jira/browse/NIFI-11158)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-11158`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-11158`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [ ] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 11
  - [x] JDK 17

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [x] Documentation formatting appears as expected in rendered files

### CODE REVIEW

**Review by Lehel44 — COMMENTED** [review]
WIll review.

File: nifi-nar-bundles/nifi-salesforce-bundle/nifi-salesforce-processors/src/main/java/org/apache/nifi/processors/salesforce/PutSalesforceObject.java [hunk 1]
```diff
@@ -170,24 +177,30 @@ public void onTrigger(ProcessContext context, ProcessSession session) throws Pro
...
                 }
             }
-
             if (writer.isActiveRecordSet()) {
                 processRecords(objectType, out, writer, extender);
             }
-            session.transfer(flowFile, REL_SUCCESS);
-
+          }
+          final long transferMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNanos);
```
Comment (Lehel44) [inline nifi-nar-bundles/nifi-salesforce-bundle/nifi-salesforce-processors/src/main/java/org/apache/nifi/processors/salesforce/PutSalesforceObject.java:184]: Could you please move this after session::transfer?

**Review by Lehel44 — CHANGES_REQUESTED** [review]
Thanks for the improvements @krisztina-zsihovszki! I looked through the changes and looks good to me. Could you please add an assertion to the new provenance event to an existing test case?

File: nifi-mock/src/main/java/org/apache/nifi/util/StandardProcessorTestRunner.java [hunk 2]
```diff
@@ -1056,4 +1058,12 @@ public void assertAllConditionsMet(final Relationship relationship, Predicate<Mo
     public void setRunSchedule(long runSchedule) {
         this.runSchedule = runSchedule;
     }
+    @Override
```
Comment (Lehel44) [inline nifi-mock/src/main/java/org/apache/nifi/util/StandardProcessorTestRunner.java:1061]: Please add new line here,

File: nifi-mock/src/main/java/org/apache/nifi/util/TestRunner.java [hunk 3]
```diff
@@ -1062,4 +1063,6 @@ void assertAttributes(
      */
      void setRunSchedule(long runSchedule);
 
+    void assertProvenanceEvent(ProvenanceEventType eventType);
```
Comment (Lehel44) [inline nifi-mock/src/main/java/org/apache/nifi/util/TestRunner.java:1066]: Please add Javadoc similar to the other methods.

File: nifi-nar-bundles/nifi-salesforce-bundle/nifi-salesforce-processors/src/main/java/org/apache/nifi/processors/salesforce/PutSalesforceObject.java [hunk 4]
```diff
@@ -170,24 +177,30 @@ public void onTrigger(ProcessContext context, ProcessSession session) throws Pro
...
             }
-
             if (writer.isActiveRecordSet()) {
                 processRecords(objectType, out, writer, extender);
             }
-            session.transfer(flowFile, REL_SUCCESS);
-
+          }
+          session.transfer(flowFile, REL_SUCCESS);
+          final long transferMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNanos);
```
Comment (Lehel44) [inline nifi-nar-bundles/nifi-salesforce-bundle/nifi-salesforce-processors/src/main/java/org/apache/nifi/processors/salesforce/PutSalesforceObject.java:185]: The whole bundle does not use the final keyword. Please remove it for consistency.

File: nifi-nar-bundles/nifi-salesforce-bundle/nifi-salesforce-processors/src/main/java/org/apache/nifi/processors/salesforce/util/SalesforceRestService.java [hunk 5]
```diff
@@ -103,6 +103,10 @@ public InputStream postRecord(String sObjectApiName, String body) {
         return request(request);
     }
 
+    public String getVersionedBaseUrl() {
+        return baseUrl + "/services/data/v" + version;
+    }
+
     private InputStream request(Request request) {
```
Comment (Lehel44) [inline nifi-nar-bundles/nifi-salesforce-bundle/nifi-salesforce-processors/src/main/java/org/apache/nifi/processors/salesforce/util/SalesforceRestService.java:110]: I recommend refactoring this method a little bit.

`private InputStream request(Request request) {
        try (Response response = httpClient.newCall(request).execute()) {
            if (response.code() < 200 || response.code() > 201) {
                throw new ProcessException("Invalid response" +
                        " Code: " + response.code() +
                        " Message: " + response.message() +
                        " Body: " + (response.body() == null ? null : response.body().string())
                );
            }
            return Objects.requireNonNull(response.body()).byteStream();
        } catch(ProcessException e) {
            throw e;
        } catch (Exception e) {
            throw new ProcessException(String.format("Salesforce HTTP request failed [%s]", request.url()), e);
        }
    }`

**Comment by turcsanyip** [comment]
@krisztina-zsihovszki Thanks for the improvements!
@mark-bathori @Lehel44 Thanks for your reviews!

Merging to main.

**Pull request closed without merging**